### PR TITLE
FM2-334:Add Intergration Test for collect-data operation in HAPI FHIR

### PR DESCRIPTION
**Pull request by gitcliff**
link:https://issues.openmrs.org/browse/FM2-334

### CODE REVIEW

**Comment by HerbertYiga** [comment]
@gitcliff  thanks cliff did you forget to add the ticket link on the description

File: src/test/java/ca/uhn/fhir/jpa/starter/OperationCollectDataTest.java [hunk 1]
```diff
@@ -142,10 +251,28 @@ private Parameters fetchParameter(String theUrl) throws IOException {
 		return parameters;
 	}
 
+	private Parameters fetchXmlParameter(String theUrl) throws IOException {
+		Parameters parameters;
+		HttpGet get = new HttpGet(theUrl);
```
Comment (mozzy11) [inline src/test/java/ca/uhn/fhir/jpa/starter/OperationCollectDataTest.java:256]: @gitcliff , you make unncesesary duplications , can you create the Get once in one private method and re-use that,with the existing tests ??

File: src/test/java/ca/uhn/fhir/jpa/starter/OperationCollectDataTest.java [hunk 2]
```diff
@@ -168,7 +295,27 @@ private void postResource(String theUrl, String filePath)
 		}
 	}
 
-	private String readJsonFile(String path) throws IOException {
+	private CloseableHttpResponse postXmlResource(String theUrl, String filePath)
+      throws IOException{
+    HttpPost post = new HttpPost(theUrl);
+    String xml = readFile(filePath);
+
```
Comment (mozzy11) [inline src/test/java/ca/uhn/fhir/jpa/starter/OperationCollectDataTest.java:302]: Also thei , Create one Post in a given privae method and re-use that with the existing tests

File: src/test/java/ca/uhn/fhir/jpa/starter/OperationCollectDataTest.java [hunk 3]
```diff
@@ -168,7 +295,27 @@ private void postResource(String theUrl, String filePath)
...
+    StringEntity entity = new StringEntity(xml);
+    post.setEntity(entity);
+    post.setHeader("Accept", "application/xml");
+    post.setHeader("Content-type", "application/xml");
```
Comment (mozzy11) [inline src/test/java/ca/uhn/fhir/jpa/starter/OperationCollectDataTest.java:306]: if   a`pplication/xml ` fails to return a json response , try  `application/fhir+xml`

File: src/test/resources/ObsBundle.xml [hunk 4]
```diff
@@ -0,0 +1,247 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Bundle xmlns="http://hl7.org/fhir">
+ <type value="collection"/>
```
Comment (mozzy11) [inline src/test/resources/ObsBundle.xml:3]: Not sure if Hapi can accept a bundle with type `collection` . it should be a `transaction`

**Comment by mozzy11** [comment]
@gitcliff ,try not to  duplicate code as much as you can , simply put the common pieces in a a method you can simply re-use

File: src/test/java/ca/uhn/fhir/jpa/starter/OperationCollectDataTest.java [hunk 5]
```diff
@@ -142,10 +270,28 @@ private Parameters fetchParameter(String theUrl) throws IOException {
 		return parameters;
 	}
 
+	private Parameters fetchXmlParameter(String theUrl) throws IOException {
```
Comment (ibacher) [inline src/test/java/ca/uhn/fhir/jpa/starter/OperationCollectDataTest.java:273]: So, I'd suggest combining `fetchParameter()` and `fetchXmlParameter()` into something like this:

```java
public static final MediaType FHIR_JSON_MEDIA_TYPE = MediaType.valueOf("application/fhir+json");
public static final MediaType FHIR_XML_MEDIA_TYPE = MediaType.valueOf("application/fhir+xml");

// ...

private Parameters fetchParameters(String theUrl, MediaTypes mediaType) throws IOException {
	HttpGet get = new HttpGet(theUrl);

	String auth = USER_NAME + ":" + USER_PASSWORD;
	byte[] encodedAuth = Base64.encodeBase64(auth.getBytes(StandardCharsets.ISO_8859_1));
	String authHeader = "Basic " + new String(encodedAuth);
	get.addHeader(HttpHeaders.AUTHORIZATION, authHeader);
	get.addHeader(HttpHeaders.ACCEPT, mediaType.toString());


	try (CloseableHttpResponse resp = ourHttpClient.execute(get)) {
		IParser parser;
		MediaType responseMediaType = MediaType.parseMediaType(response.getContentType());
        if (FHIR_JSON_MEDIA_TYPE.isCompatibleWith(responseMediaType) || MediaType.APPLICATION_JSON.isCompatibleWith(responseMediaType)) {
			parser = ourCtx.newJsonParser();
		} else if (FHIRXML_MEDIA_TYPE.isCompatibleWith(responseMediaType) || MediaType.APPLICATION_XML.isCompatibleWith(responseMediaType) || MediaType.TEXT_XML.isCompatibleWith(responseMediaType)) {
			parser = ourCtx.newXmlParser();
		} else {
			throw AssertionError("Cannot handle response of type " + responseMediaType.toString());
		}

		return parser.parseResource(Parameters.class,
			  EntityUtils.toString(resp.getEntity(), Charsets.UTF_8));
	}
	
	throw new AssertionError("Could not retrieve parameters from server");
}
```

Then use just changes to:

```java
fetchParameters(theUrl, FHIR_JSON_MEDIA_TYPE);
// or
fetchParameters(theUrl, FHIR_XML_MEDIA_TYPE);
```

I'd do something similar with `postResource()` and `postXmlResource()`

File: src/test/resources/FhirMeasure.xml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+<status value="draft"/> 
+<experimental value="true"/> 
+
+  <group id="population">
+    <population> 
+      <code> <coding> <code value="initial-population"/> </coding> </code> 
+      <criteria> <language value="application/x-fhir-query"/> <expression value="Observation?code=856AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA,1305AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA&amp;_include=Observation:patient"/> </criteria> 
+      </population>
+  </group>    
+</Measure>
```
Comment (ibacher) [inline src/test/resources/FhirMeasure.xml:19]:
```suggestion
</Measure>

```

**Review by ibacher — CHANGES_REQUESTED** [review]
@gitcliff Thanks for this! I've got a few suggestions and a couple of changes to make.

File: src/test/java/ca/uhn/fhir/jpa/starter/OperationCollectDataTest.java [hunk 7]
```diff
@@ -66,16 +80,24 @@
...
 
+	private Parameters parameters;
+
+	private String paramName1 = "measureReport";
+
+	private String paramName2 = "resource";
+	
+	private HttpGet get;
+
+	private HttpPost post;
```
[the remark this answers is not in the document]
Comment (gitcliff) [reply src/test/java/ca/uhn/fhir/jpa/starter/OperationCollectDataTest.java]: sure, i have fixed it, 
kindly review

File: src/test/java/ca/uhn/fhir/jpa/starter/OperationCollectDataTest.java [hunk 8]
```diff
@@ -125,33 +280,58 @@ void beforeEach() {
...
+				parser = ourCtx.newJsonParser();
+			} else if (FHIR_XML_MEDIA_TYPE.isCompatibleWith(responseMediaType) || MediaType.APPLICATION_XML.isCompatibleWith(responseMediaType) || MediaType.TEXT_XML.isCompatibleWith(responseMediaType)) {
+				parser = ourCtx.newXmlParser();
+			}
+			 else {
+				throw new AssertionError("Cannot handle response of type " + responseMediaType.toString());
+			}
+	
+			return parser.parseResource(Parameters.class,
+				  EntityUtils.toString(resp.getEntity(), Charsets.ISO_8859_1));
```
Comment (ibacher) [inline src/test/java/ca/uhn/fhir/jpa/starter/OperationCollectDataTest.java:307]: We should be using `Charsets.UTF_8` as this is the default encoding for both XML and JSON files. Roughly speaking UTF-8 is a superset of ISO 8859-1.

Comment (gitcliff) [reply src/test/java/ca/uhn/fhir/jpa/starter/OperationCollectDataTest.java]: I have fixed it

**Review by ibacher — APPROVED** [review]

**Comment by ibacher** [comment]
@mozzy11 Are you ok with merging this?

**Review by mozzy11 — APPROVED** [review]

**Pull request merged**